File: native/cpp/kernel.cpp
#include "kernel.h"

#include <algorithm>
#include <cmath>
#include <sstream>

namespace void_native {

const char* kernel_name() {
	return "VOID native cognitive kernel";
}

unsigned int kernel_version() {
	return 2;
}

unsigned int kernel_abi_version() {
	return 1;
}

KernelInfo kernel_info() {
	return {kernel_name(), kernel_version(), kernel_abi_version(), true};
}

bool supports_dimensions(unsigned int dimensions) {
	return dimensions > 0 && dimensions <= 4096;
}

bool supports_batch(unsigned int batch_size) {
	return batch_size > 0 && batch_size <= 4096;
}

double clamp_unit(double value) {
	if (!std::isfinite(value)) return 0.0;
	return std::clamp(value, 0.0, 1.0);
}

double safe_divide(double numerator, double denominator) {
	if (!std::isfinite(numerator) || !std::isfinite(denominator) || denominator == 0.0) return 0.0;
	return numerator / denominator;
}

std::string kernel_diagnostics(unsigned int dimensions, unsigned int batch_size) {
	const KernelInfo info = kernel_info();
	std::ostringstream result;
	result << info.name << " v" << info.version
		   << " abi=" << info.abi
		   << " dimensions=" << dimensions
		   << " batch=" << batch_size
		   << " valid=" << (supports_dimensions(dimensions) && supports_batch(batch_size));
	return result.str();
}

}  // namespace void_native

File: native/rust/src/kernel.rs
pub const NAME: &str = "VOID native cognitive kernel";
pub const VERSION: u32 = 2;
pub const ABI_VERSION: u32 = 1;
pub const MAX_DIMENSIONS: usize = 4096;
pub const MAX_BATCH: usize = 4096;

pub fn signature() -> String { format!("{} v{} abi={}", NAME, VERSION, ABI_VERSION) }

pub fn supports_dimensions(dimensions: usize) -> bool { dimensions > 0 && dimensions <= MAX_DIMENSIONS }

pub fn supports_batch(batch: usize) -> bool { batch > 0 && batch <= MAX_BATCH }

pub fn clamp_unit(value: f64) -> f64 {
	if !value.is_finite() { return 0.0; }
	value.clamp(0.0, 1.0)
}

pub fn safe_divide(numerator: f64, denominator: f64) -> f64 {
	if !numerator.is_finite() || !denominator.is_finite() || denominator == 0.0 { return 0.0; }
	numerator / denominator
}

pub fn diagnostics(dimensions: usize, batch: usize) -> String {
	format!("{} dimensions={} batch={} valid={}", signature(), dimensions, batch, supports_dimensions(dimensions) && supports_batch(batch))
}
